### ISSUE TITLE
Investigate Image Intervention for better image handling

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1472]
https://github.com/Intervention/image

https://forum.textpattern.com/viewtopic.php?id=50612
https://forum.textpattern.com/viewtopic.php?pid=321781#p321781

**Comment by philwareham** [comment]
Nice find! I'd love to see better image handling in Textpattern.

**Comment by petecooper** [comment]
Tagging https://github.com/textpattern/textpattern/issues/913 for completeness.

**Comment by philwareham** [comment]
@bloke you interested in helping me with an update to the image handling in Textpattern for 4.9? Admittedly all I can provide is UI and UX but I'd be very keen on seeing this move forward - with or without the library that Pete suggests (hopefully with, if it is workable).

**Comment by petecooper** [comment]
I can provide a workable image corpus, if that helps.

**Comment by philwareham** [comment]
We still have this long-term goal to integrate too, which this library could help with (and the corresponding mini grid version on the write page): https://design-patterns.textpattern.com/mockups/content-image-grid.html

**Comment by Bloke** [comment]
Definitely. This has been far too long mooching in the shadows. I'll dust off the image doc and see which bits are easily achievable. I'd really like to work towards pulling in better core support for `srcset` / `picture`

I'm assessing the whole concept of "what is a thumbnail" and "when is a thumbnail not a thumbnail" to figure out the best route forward. smd_thumbnail is partway there, but it's too unwieldy for general use and there are some trivially stupid cases it can't handle and should, that we need to figure out before we can boil it down to something lean and core worthy.

**Comment by philwareham** [comment]
Assessing the library Pete mentioned would be good I think - it's been in active development for over 7 years with regular releases and thorough documentation, so I think it's around to stay. That would hopefully take some of the hard work out of achieving our goal.

**Comment by philwareham** [comment]
Also, the license is compatible (MIT licensed works can be bundled with GNU General Public License v2.0 without conflict).

**Comment by petecooper** [comment]
Note: Intervention required FileInfo, which has been on by default since PHP 5.3: https://www.php.net/manual/en/fileinfo.installation.php - would that warrant a sys reqs tweak at our end?

**Comment by philwareham** [comment]
Does this also give us WebP support in PHP5.5? Seems like it from the documentation http://image.intervention.io/getting_started/formats

**Comment by petecooper** [comment]
@philwareham - if we get this into 4.9, I'm guessing we'll bump the sys reqs to PHP 5.6. I would hope in a year many or most web hosts will be well into PHP 7 and PHP 5 will be an also-ran. I hope.

**Comment by petecooper** [comment]
Note: Image Intervention has a minimum version requirement for GD / ImageMagick, so we may need to be more specific on our sys reqs.

**Comment by philwareham** [comment]
Related too: https://github.com/ambroisemaupate/intervention-request

**Comment by petecooper** [comment]
Tagging https://github.com/thephpleague/glide so we don't forget.

**Comment by philwareham** [comment]
And https://github.com/thephpleague/flysystem (which is included along with Intervention Image when you `composer require league/glide`.

**Comment by philwareham** [comment]
Just reinstalled WordPress locally for some reference as to the market leader does it. Editing tools are fairly basic and you can undo prior to save if things go wrong. The filesystem does store the original upload intouched and then uses a 2560px version as the basis for any cropping and other destructive editing. Then a range of sizes scaled down from that - although I can't see exactly where all these profiles are explicitly set (the settings -> media section only mentions three sizes, so not sure where it deems to gets the other sizes from).

![Screenshot 1](https://user-images.githubusercontent.com/413665/89291145-e4f24880-d651-11ea-9c67-123fe4894c1a.png)

![Screenshot 2](https://user-images.githubusercontent.com/413665/89291148-e754a280-d651-11ea-89bc-3bb4ae7d562b.png)

**Comment by philwareham** [comment]
Essential tools from Intervention Image:

- Crop
- Flip horizontal
- Flip vertical
- Rotate left/right

Possible handy extras (maybe not core, but exposed via a plugin)

- Greyscale
- Brightness
- Contrast
- Blur
- Sharpen
- Gamma
- Colorize

If we can get this working, we could drag Textpattern from prehistoric image handling to very powerful in one fell swoop. Very exciting. I can play around with a suitable UI for the above in due course. Our current theme UI has most of what we need already, on that front.

**Comment by philwareham** [comment]
I started playing with some rudimentary layout for an image edit pane: https://design-patterns.textpattern.com/mockups/secondary/content-image_edit-proposal.html

Yeah, it's a rough WIP but you get the idea of where we could go.

**Comment by Bloke** [comment]
Scrummy!

**Comment by Bloke** [comment]
Depending on where we go with creating various sizes, we might not even need the thumbnail management area between the caption and Save/GoBack buttons, which saves us a little more real estate.

Maybe once we figure out Intervention's capabilities, we can simply stack the various versions of each image that have been created/cached so far at the bottom, with an indication of which size each is? Maybe we could have the option to delete (flush it from the cache) and/or replace with a dedicated image - if Intervention makes it easy for us to do this.

**Comment by bloatware** [comment]
We might also delegate caption (and even alt?) to somewhere else. Why an image could not be used with two different captions?

**Comment by Bloke** [comment]
Generally, one caption and one alt will cover all versions of an image. But in the cases where you've replaced an individual image version (say, the 3rd one because at that resolution it offers a zoomed-in view to show detail) then it might make sense to be able to supply a different alt/caption just for that one image, which would override the default one on the master.

The problem we have is that all these 'image instances' beyond the 'main' (uploaded) version don't have any metadata. They don't have a database entry so where do we store this custom caption? We could keep a parallel table with the sizes that have been generated and a pointer back to the original image ID. In that secondary table, there could be columns for alt and caption and maybe even sizes, etc.

However, if we go the route of auto-generating images when they're used on the front-end, we'd have start adding database entries each time an image was viewed on the front-end, which is a bunch of write operations that make me a little nervous. If we can make the insertions safe, then maybe.

Once an image is generated and has a cache file, then you could go into the Image Edit panel and find the relevant pic to supply its additional metadata.

Dunno. The beauty of not having this stuff is that the cached versions pretty much manage themselves. But if we start to hold additional metadata about them it adds a layer of complexity.

I'd be keen to find alternative ways to store this stuff before we introduced database writes for each image that's viewed.

**Comment by philwareham** [comment]
My opinion is that one image entry (and its set if sizes) has one metadata for all. Not separate entries per image. These image sets are primarily to serve correct images in srcsets.
I’m also unsure of whether autogenerating cached images is viable, as opposed to defining a set of sizes in image prefs. I’d like to be proved wrong but I can’t be sure.

**Comment by petecooper** [comment]
I love it when a plan comes together.

**Comment by Bloke** [comment]
i concur.

Thinking about it more while I was vacuuming (I love mindless tasks to free think!) one image = one metadata record is simplest and the right thing to do. It's most likely to be what people use. If anyone needs to do 'zoomed in' or 'alternative' versions of a single `<picture>` element, they should upload a different image and create any alternate sizes of that image they want.

All we have to do is make it easy for people to use tags to pluck out multiple images where they want them. That should be fairly simple if we supercharge the `<txp:image>` tag too. We could add `media`, `sizes` and `type` attributes (with valueless `type` returning the media type of the image itself so you'd only need to override it if you wished to do something crazy). Starting simple:

``` html
<txp:images name="kitten-stretching.png, kitten-sitting.png, kitten-curled.png"
      wraptag="picture"
      media="(min-width: 650px), (min-width: 465px)">
   <txp:if_last_image>
      <txp:image />
   <txp:else />
      <source srcset="<txp:image_url />" <txp:image_info type="media" /> />
   </txp:if_last_image>
</txp:images>
```

Tada! Art direction:

``` html
<picture>
  <source 
    media="(min-width: 650px)"
    srcset="images/kitten-stretching.png">
  <source 
    media="(min-width: 465px)"
    srcset="images/kitten-sitting.png">
  <img 
    src="images/kitten-curled.png" 
    alt="a cute kitten">
</picture>
```

Not sure how we'd specify multiple images for a single scrset or 'x' densities, but it's just hashing through some examples and coming up with the best syntax.

**Comment by Bloke** [comment]
Further thinking on the predefined vs autogenerate scenarios. How about both?

On the image list panel where we currently add the category prior to upload, add a box that accepts a comma-separated list of pixel widths. If it's empty, you get just the one main image uploaded. If it's got one or more numbers in the box, then it generates images from the main one at those widths on upload. That would apply to _all_ images added in that set allowing you to batch create smaller versions at whatever sizes you want.

When you upload, the values you supplied in the 'widths' box is stored as a hidden pref so it's remembered for you next time. Means you can effectively do it once and leave it, then every time you upload images, you'll get a bunch made to the sizes you used last time. You can alter this at any time and the new value is stored after upload.

Most people will be happy at that. The sizes can be used as the designer sees fit, pre-computed to their sizes to use in their sites. So, no need for them to be generated on-the-fly as we've generated them on-the-fly... in advance!

Here's the kicker. Internally, when they ask for widths 300, 600, 1200 and hit upload, for each main image that's uploaded, we pass it through an API (via cURL) that we design for creating an image at each given dimension/crop/whatever. That interfaces with Image Intervention to create the various sizes.

Now, what if this API is _also_ available to the various image tags? So in the event that you request an image size that you haven't already pre-computed, it'll be created and cached wherever you've told it to store them via prefs (local filesystem, cloud, whatever). That's a nice fallback.

On the Images list panel, it's the same deal. When you upload a bunch of images, the lazy loader will try to fetch the thumbs to display (list or grid sizes, depending on what is currently showing) from the API URL. If the person has constructed them at the size we want for that view, great, they're already made and we show them. If not, they're made at that moment, and sit alongside the ones they defined at their designated endpoint (local filesystem, cloud, whatever).

Result: we see the thumbs and they are clickable to jump to the Edit panel.

Speaking of which, when you click through to Phil's revamped Image Edit panel, how about (if we can) beneath the main image, add a dropdown that lists all the sizes of image available - i.e. all the ones created in the cache so far for that image. The top, default, entry is 'Full size' (or whatever we designate it) so you meddle with that using the contrast/brightness/crop/rotate/whatever tools.

If, however, you pick another size from the list, _that image is loaded into the panel instead_. Thus you can crop, rotate, color, tint, scale, whatever _that image_. When you save, it's replaced at the designated endpoint.

That allows you to mass create images on upload and then go in and tweak them using the swanky UI if you wish.

Further, on the Edit panel:

* Add a checkbox (near the save button? Depends if the edits are committed prior to saving or the edited clone is manipulated on the fly) labelled "Recreate all" so that when you alter the original - and thus create this parallel 'edited' version - when you commit the final changes and leave the panel, all your current thumbs for that image are recreated from the modified version. That way, you can choose to just edit the main (which regenerates only the edited clone) and leaves the others alone, in case you've manually mucked with them and don't want them trashed, for whatever reason.
* Add a 'width' text entry box alongside the current dropdown of available sizes and an 'Add' button nearby. So if you see one or two sizes missing and want to pre-create them, you can just type the widths in the box and hit Add. They're passed through the API via cURL and the given width images are immediately made for you.
* Bonus: an 'x' button allows you to delete the currently selected sized image from the dropdown - the one currently showing in the interface beneath the new editing tools Phil added to the mockup. Maybe the main image cannot be deleted this way and can only be removed from the multi-edit tool? Not sure.
* Plugins can intervene in the image creation URL so tweaks can be made. Maybe new attributes we haven't even thought of could be added via a plugin? And plugins such as com_article_image can of course call the API to create thumbs at any given widths if the widths are exposed via a box in its own UI. Or just not bother and rely on the lazy load/first use of an image to create the thumbs on behalf of the system.

That sounds super workable to me. A great mix of auto-generated and pre-computed. Some fabulous new tools to operate on one or more of the images. The ability to offload them to remote storage. Image effects, manipulation, and the ability for plugins to extend whatever we design.

And _that_ is why you pay me the big bucks :D

**Comment by bloatware** [comment]
I just meant *caption* is not metadata. An image could have `My father` caption in some context, and `My brother` in a different one.

**Comment by Bloke** [comment]
> I just meant _caption_ is not metadata. An image could have `My father` caption in some context, and `My brother` in a different one.

Eh? If you know a particular image (are we talking variant of a main image here at a different width?) is of something else or needs captioning differently, then override the stored caption at the tag level? Or am I being super dense here?

**Comment by bloatware** [comment]
Sure, we can override the caption, which means (for me) it's not inherent to the image itself, unlike the author, the creation date and other metadata. But 'default' captions are handy, I agree.

**Comment by philwareham** [comment]
I've done a bit more work on the panel visual now: https://design-patterns.textpattern.com/mockups/secondary/content-image_edit-proposal.html

**Comment by petecooper** [comment]
FYI, when the CSP stuff is untangled (great work on that, by the way), let me know if you want a demo spun up for the imagey stuff - happy to help where I can.

**Comment by petecooper** [comment]
Tagging https://github.com/textpattern/textpattern/issues/1609

**Comment by jools-r** [comment]
Bit late to the show here, but could https://github.com/spatie/image and https://github.com/spatie/image-optimizer be of relevance here. Uses [glide](http://glide.thephpleague.com) under the hood and is kept up to date.

**Comment by philwareham** [comment]
@jools-r possibly, but I'm fairly happy with the original suggestion. It's also used by Statamic at least, so continuous development is fairly guaranteed, and fulfils our criteria.

**Comment by philwareham** [comment]
Further considerations when we revamp the image handling. We should also cater for different image formats per entry too.

So, for example, you have uploaded your 'hero' image `23.png`. You also want to allow a `23.webp` and/or `23.avif` and/or `23.jxl`etc. to exist for that image. This can also be useful with our new  `<txp:image_info type="mime" />` that landed with Textpattern 4.8.5.

Note: I don't necessarily want the system to generate those images (although if it could, marvellous), but rather a way to upload them in the control panel and/or for the CMS to know they exist. This keeps the image manipulation to a minimum since we are not existing the library to generate cutting-edge image formats like AVIF and JPEG XL on top of what we've already discussed above.

My current (4.8.5) workaround for this is:

1. Upload one of the '100% supported' 2021 image types, i.e. PNG, JPEG, GIF (e.g. `24.jpg`).
2. Use a third-party tool to generate other image file formats (my current choice is https://github.com/GoogleChromeLabs/squoosh/ - which I've just discovered has a CLI version too 🎉).
3. FTP upload those extra images to Textpattern's images directory with same ID name as what I did in step 1 (e.g. `23.webp`).
4. Use a `<picture>` tag with some Textpattern tags sprinkled within it to generate a modern image solution.

Drawbacks of that workaround are:

1. I have to manually FTP into the images directory (instead of uploading somehow within the Textpattern admin panel).
2. The CMS doesn't know which extra file formats I've uploaded and are available (this isn't an issue for sites I manage directly as I can ensure the correct image formats are always uploaded).

Sidenote to this, in Textpattern 4.9 we need to extended the image MIME types for images with:

- AVIF [caniuse](https://caniuse.com/?search=AVIF)
- JPEG XL [caniuse](https://caniuse.com/?search=jpeg%20xl)
- WebP2 (TBC, currently in development)

**Comment by petecooper** [comment]
`libgd` v2.3.2 has landed, which includes support for AVIF (via `libavif`) and HEIF (via `libheif`).

**Comment by philwareham** [comment]
> `libgd` v2.3.2 has landed, which includes support for AVIF (via `libavif`) and HEIF (via `libheif`).

Good to know, thanks. I don't really think HEIF is going to win out as a web image format due to the patent issues, and that AVIF is essentially the same but more open - but good to know it is at least supported at server-level. JPEG XL is potentially the best longterm image format but fingers crossed it gains widespread support.

AVIF really seems to flatten out minor detail, especially in skin tones, at mid-level compression settings in my tests. There is no easy 'one default compression ratio fits all' for that formats whereas if you go around 50% compression on WebP and JPEG you can almost guarantee a good balance between quality and compression (apart from in reds, that JPEG always seems to struggle with artefacts on).

**Comment by petecooper** [comment]
@philwareham - you should be a designer, you'd probably be OK at that, I reckon.

**Comment by philwareham** [comment]
Ha! I actually prefer doing design for print but the internet keeps dragging me back in.

**Comment by Bloke** [comment]
> I don't really think HEIF is going to win out as a web image format due to the patent issues, and that AVIF is essentially the same but more open... JPEG XL is potentially the best long-term image format but fingers crossed it gains widespread support.

VHS/Betamax
HD DVD/Blu-Ray
...

**Comment by petecooper** [comment]
I suppose if we can make the image handing more modular and accept there is some disparity between browser support, we could slap a fat _caveat emptor_ on things and let admins run with scissors if they wish.

**Comment by Bloke** [comment]
Yep. The good thing with using an external library for handling them is that, as long as the library is kept up-to-date, it's easier for us to keep up with the latest happenings in image format land. In fact, if we're careful to make an internal Venn diagram of the union between what the server and library support, Txp should benefit without us having to do much extra work at all. Update external library from time to time. Done.

**Comment by philwareham** [comment]
Yes my feeling is we directly support the image formats in Image Intervention (which itself takes its support from what's in GD or ImageMagick, so as Stef says very little extra work on our part after the library is in place) and then **also** allow you a way of uploading other formats of the same image that currently aren't supported by Image Intervention (and which won't directly be manipulated by Image Intervention).

So, basically the process would be:

1. You upload your original primary source image to Textpattern in your primary format (say JPEG for example).
2. You manipulate that image to the size you want, any optional filters, whatever.
3. You generate the primary website-ready image. Or set of website-ready images at various sizes if doing responsive images.
4. You have a field where you can designate that another format of this image exists (either another image ID in the Textpattern system or a simple text field where you can supply a comma separated list of MIME types or a set of text field and file upload combinations where you can upload an image and also designate it's MIME - not really sure what the best route would be yet).
5. Build you `picture` tag or `img srcset` tag

Quite a lot to take in, but I think the first step on this path is to get 1 - 3 sorted. Adding other image formats in would be something to work out afterwards (and maybe for even 4.10 release).

**Comment by philwareham** [comment]
There's some functionality in smd_thumbnail that also helps with ideas for point 3 above (the generation of multiple sizes) - that works fairly logically in my opinion. Unless Image Intervention can do that natively.

**Comment by philwareham** [comment]
Also to bear in mind, we need the initial process of uploading an image and using it to still be as simple as possible. We just allow more complexity if the user wishes to harness it. So you could still just upload a single image and do no further manipulation of it and still add to your site.

**Comment by phiw13** [comment]
> Also to bear in mind, we need the initial process of uploading an image and using it to still be as simple as possible.

Can we also have the same process as what is possible with smd_thumbnail today - manually uploading images for the (various) profiles and not depending on whatever lib is available?

[I did some tests today: manually created webP from Tiff file vs whatever `gd` did to size it (based on the large size uploaded WebP image): 20Kb for the first one vs 32Kb for the second. Exact same pixel size.]

**Comment by phiw13** [comment]
>  In fact I'd be pretty happy if the functionality of smd_thumbnail was brought over wholesale into the core (with some tag naming adjustments)

+ 10 with some nice serving of spring sunlight !

**Comment by Bloke** [comment]
@philwareham Do you have the location of the current 4.9 mockups please? The links above 404 :(

I've found: https://design-patterns.textpattern.com/stable/mockups/ but I'm not sure they're current. I swear I saw a snazzy mockup with tint/greyscale/etc buttons on it. https://design-patterns.textpattern.com/dev/mockups/ seem to be the same. Ta!

**Comment by philwareham** [comment]
Sure, it's been moved here: https://design-patterns.textpattern.com/dev/mockups/secondary/content-image_edit-proposal.html

**Comment by Bloke** [comment]
Ahhhh, thank you. Threw me because the -proposal file's not listed in the index page.

**Comment by petecooper** [comment]
JPEG XL support just got binned from Chromium:

https://chromium-review.googlesource.com/c/chromium/src/+/4081749
https://news.ycombinator.com/item?id=33933208

Associated bug:

https://bugs.chromium.org/p/chromium/issues/detail?id=1178058

**Comment by philwareham** [comment]
Well that's another shit decision by Google. JPEG XL is far and away better than AVIF, and don't get he started on the terrible WebP format. Politics.

**Comment by Bloke** [comment]
I'm still torn here. Part of me wants to simply enable some additional post-processing facility in core and leave it to plugins to implement the detail. I could see com_image_edit, for example, providing the ability to flip, rotate, sharpen, blur, grayscale, etc. So if people want this facility, it's a single, project-sanctioned, plugin away, but those that don't need that facility can stick with what we have.

The key driver, imo, is to handle resizing elegantly in core. Possibly cropping too. That has multiple components:

1. A way to resize on-demand (like SLIR and others). Ability to limit the sizes that are available, or sign the requests, so image creation can't be tampered with is a must.
2. A way to cache generated images. With garbage collection.
3. A way to intelligently spread the load (on disk) of the generated images so they're not all lumped into a single directory.

SLIR caters for the first two. Sort of. Other libraries (e.g. Glide, Image, Imagine) tend to go for the image processing route and have resizing as a side gig but don't handle cacheing natively. Glide has bonus points insofar as it uses FlySystem which enables offloading of image storage to basically anywhere on the cloud, meaning vast Amazon S3 image libraries are doable, with local cacheing for speed.

Almost all of the latter, however, are based on Intervention, or Glide. Which are Laravel-based things that come with a tonne of dependencies. That bumps our core by, at minimum for Intervention, 750KB. More for any libraries built on top of it. And we won't use much of it.

I thought that [Imagine](https://github.com/php-imagine/Imagine) might be quite good as it doesn't require Intervention at all, but claims to offer a unified mechanism for accessing GD, Imagick and GMagick inside PHP. Reading the docs, though, you don't just access one object and call methods on it, you have to instantiate one of the three and _then_ call the methods. And not all the methods are available for all the libraries. That complicates matters. Sure, we could put a wrapper around it that checks what's installed and instantiates one of them and maybe we intnerally keep a note of what is supported by what so we can expose only the relevant functionality to core/plugins. But it's extra faff. Also, Imagine is 8.5MB! However, 8MB of that are Adobe .icc profiles. If there was a way to include Imagine without_ those and just bypass the colour profiles, that might be doable, but it seems to be embedded into the fabric of the library.

So I think Imagine is, unfortunately, not a contender. I'm not willing to more than double our codebase just for the option of manipulating images, and then have to build our own cacheing on top.

Unless anybody can come up with a viable library that's small, efficient and scalable that fits our needs above, it looks like we might have to pick and choose the best bits of all the tools mentioned above, and roll our own implementation. Yeah yeah, Not Invented Here. But I really can't see much choice, given they're all either 'kitchen sink' approaches, too big, too limited, not scalable, require additional code, require forking and maintaining, or have PHP version requirements that outstrip our minimum.

If anyone has any thoughts on this or can find one (or more) tiny libraries that meet our needs, I'm seriously tempted to write one.
